[iOS] Dispatch active "touchmove" events without sending synchronous IPC messages

whsieh opened the pull request:
#### b7a998fb59f59222ea76fe60141f38cfd2fc8cde
<pre>
[iOS] Dispatch active &quot;touchmove&quot; events without sending synchronous IPC messages
<a href="https://bugs.webkit.org/show_bug.cgi?id=243250">https://bugs.webkit.org/show_bug.cgi?id=243250</a>
rdar://78840792

Reviewed by Tim Horton and Devin Rousso.

Dispatch active &quot;touchmove&quot; events asynchronously, by introducing a new deferring gesture recognizer
that defers pan and pinch gesture recognition until after the first &quot;touchmove&quot; event has been
handled by the page.

Compatibility-wise, this still allows pages to call `preventDefault()` on &quot;touchmove&quot; events in
order to prevent scrolling and pinch-zooming, as long as the page (at least) prevents the initial
touchmove event. This roughly matches shipping behavior, but with the slight risk that in shipping
WebKit, calling prevent default on the second &quot;touchmove&quot; event *sometimes* prevents scrolling
(whereas after this patch, preventing the second &quot;touchmove&quot; event but not the first does not
prevent native gestures). It&apos;s important to note that this behavior is already inconsistent in
shipping software; before and after this patch, the only reliable way to prevent scrolling on
&quot;touchmove&quot; events is to (at least) prevent the first &quot;touchmove&quot; event that&apos;s dispatched when
performing a pan gesture.

See comments below for more details.

Tests:  fast/events/touch/ios/prevent-touchmove-with-slow-event-listener-prevents-scrolling.html
        fast/events/touch/ios/slow-touchmove-event-listener-does-not-prevent-scrolling.html

* LayoutTests/fast/events/touch/ios/prevent-touchmove-with-slow-event-listener-prevents-scrolling-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/prevent-touchmove-with-slow-event-listener-prevents-scrolling.html: Added.
* LayoutTests/fast/events/touch/ios/slow-touchmove-event-listener-does-not-prevent-scrolling-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/slow-touchmove-event-listener-does-not-prevent-scrolling.html: Added.

Add a couple of tests to exercise cases where a page installs an active &quot;touchmove&quot; event listener
that spins for a significant amount of time (in this case, 100 ms), and then proceeds to either
prevent default (i.e. the former test), or not prevent default (the latter). In particular, this
exercises the ability for the new &quot;touchmove&quot; deferring gesture recognizer to prevent scroll view
pan gestures from firing, until the web page has at least handled the first &quot;touchmove&quot; event.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::handlePreventableTouchEvent):

Remove the codepath for dispatching sync touch events altogether, now that all preventable (active)
touch events no longer require synchronous dispatch in order to prevent default gestures from
recognizing. We introduce and use the `m_touchMovePreventionState` flag here to know when the first
preventable &quot;touchmove&quot; is being handled by the web page, and only prevent native gestures from
firing by &quot;lifting&quot; the gate for deferred &quot;touchmove&quot; gestures for the initial &quot;touchmove&quot;. This is
covered by the new test: `prevent-touchmove-with-slow-event-listener-prevents-scrolling.html`.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::doneDeferringTouchMove):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Add support for a new deferring gesture recognizer which prevents all pan and pinch gestures at (or
underneath) the web view from firing, until the web page has had a chance to prevent active
&quot;touchmove&quot; events. Note that we&apos;re able to forgo the need for separate &quot;immediately resettable&quot;
and &quot;delayed resettable&quot; deferring gestures here, since the only native gestures that can be
prevented by preventing &quot;touchmove&quot; events are pan and pinch gestures, which immediately reset in
the first place.

(-[WKContentView setUpInteraction]):
(-[WKContentView _webTouchEventsRecognized:]):
(-[WKContentView deferringGestures]):
(-[WKContentView _doneDeferringTouchMove:]):
(-[WKContentView deferringGestureRecognizer:didEndTouchesWithEvent:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::cancelCurrentInteractionInformationRequest):
(WebKit::WebPage::cancelGesturesBlockedOnSynchronousReplies): Deleted.
(WebKit::WebPage::touchEventSync): Deleted.

Delete more code (and hacks) around synchronous touch event dispatch.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/252909@main">https://commits.webkit.org/252909@main</a>
</pre>

hortont424 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -984,6 +984,10 @@ - (void)setUpInteraction
     _touchEndDeferringGestureRecognizerForSyntheticTapGestures = adoptNS([[WKDeferringGestureRecognizer alloc] initWithDeferringGestureDelegate:self]);
     [_touchEndDeferringGestureRecognizerForSyntheticTapGestures setName:@"Deferrer for touch end (synthetic tap)"];
 
+    // FIXME: Do I need more?
```
Do... you???

whsieh replied:
(OOPS!) As it turned out, I did not! Will remove.

(I forgot to remove that FIXME, which was in reference to _potentially_ requiring another gesture subgraph split by using 2 touchmove deferring gestures.)

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -8536,16 +8557,15 @@ - (BOOL)deferringGestureRecognizer:(WKDeferringGestureRecognizer *)deferringGest
-        if ([gesture isKindOfClass:UITapGestureRecognizer.class]) {
-            UITapGestureRecognizer *tapGesture = (UITapGestureRecognizer *)gesture;
+        if ([gestureRecognizer isKindOfClass:UITapGestureRecognizer.class]) {
+            UITapGestureRecognizer *tapGesture = (UITapGestureRecognizer *)gestureRecognizer;
```
Might be nice to use `dynamic_objc_cast`.

whsieh replied:
Will do!

dcrousso commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3254,65 +3257,62 @@ void WebPageProxy::handlePreventableTouchEvent(NativeWebTouchEvent& event)
+        bool didFinishDeferringTouchMove = false;
+        ASSERT_IMPLIES(event.type() == WebEvent::TouchMove, m_handlingPreventableTouchMoveCount);
+        if (event.type() == WebEvent::TouchMove && m_handlingPreventableTouchMoveCount) {
+            m_hasPreventedTouchMove |= handled;
```
Should we maybe make `m_hasPreventedTouchMove` into an `std::optional<bool>` so that we only ever set it on the first `"touchmove"`?  I think it might be better to have things be (slightly) more predictable and (slightly) list racey.

whsieh replied:
I suppose we can....try. Not sure if it's worth the risk, though.

(Also we agreed to use an `enum class` instead of `std::optional<bool>`, for better readability).

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -1795,18 +1798,23 @@ - (void)_webTouchEventsRecognized:(UIWebTouchEventsGestureRecognizer *)gestureRe
+        }
+
+        if (!_page->isHandlingPreventableTouchMove())
+            stopDeferringNativeGesturesIfNeeded(_touchMoveDeferringGestureRecognizer.get());
```
NIT: any reason this cant be moved in between the stuff for `"touchstart"` and the stuff for `"touchend"`?

whsieh replied:
No reason one way or another — it should not make a difference.

dcrousso reviewed (approved):
aaawwweeesssooommmeee

whsieh committed:
[iOS] Dispatch active "touchmove" events without sending synchronous IPC messages
https://bugs.webkit.org/show_bug.cgi?id=243250
rdar://78840792

Reviewed by Tim Horton and Devin Rousso.

Dispatch active "touchmove" events asynchronously, by introducing a new deferring gesture recognizer
that defers pan and pinch gesture recognition until after the first "touchmove" event has been
handled by the page.

Compatibility-wise, this still allows pages to call `preventDefault()` on "touchmove" events in
order to prevent scrolling and pinch-zooming, as long as the page (at least) prevents the initial
touchmove event. This roughly matches shipping behavior, but with the slight risk that in shipping
WebKit, calling prevent default on the second "touchmove" event *sometimes* prevents scrolling
(whereas after this patch, preventing the second "touchmove" event but not the first does not
prevent native gestures). It's important to note that this behavior is already inconsistent in
shipping software; before and after this patch, the only reliable way to prevent scrolling on
"touchmove" events is to (at least) prevent the first "touchmove" event that's dispatched when
performing a pan gesture.

See comments below for more details.

Tests:  fast/events/touch/ios/prevent-touchmove-with-slow-event-listener-prevents-scrolling.html
        fast/events/touch/ios/slow-touchmove-event-listener-does-not-prevent-scrolling.html

* LayoutTests/fast/events/touch/ios/prevent-touchmove-with-slow-event-listener-prevents-scrolling-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/prevent-touchmove-with-slow-event-listener-prevents-scrolling.html: Added.
* LayoutTests/fast/events/touch/ios/slow-touchmove-event-listener-does-not-prevent-scrolling-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/slow-touchmove-event-listener-does-not-prevent-scrolling.html: Added.

Add a couple of tests to exercise cases where a page installs an active "touchmove" event listener
that spins for a significant amount of time (in this case, 100 ms), and then proceeds to either
prevent default (i.e. the former test), or not prevent default (the latter). In particular, this
exercises the ability for the new "touchmove" deferring gesture recognizer to prevent scroll view
pan gestures from firing, until the web page has at least handled the first "touchmove" event.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::handlePreventableTouchEvent):

Remove the codepath for dispatching sync touch events altogether, now that all preventable (active)
touch events no longer require synchronous dispatch in order to prevent default gestures from
recognizing. We introduce and use the `m_touchMovePreventionState` flag here to know when the first
preventable "touchmove" is being handled by the web page, and only prevent native gestures from
firing by "lifting" the gate for deferred "touchmove" gestures for the initial "touchmove". This is
covered by the new test: `prevent-touchmove-with-slow-event-listener-prevents-scrolling.html`.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::doneDeferringTouchMove):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Add support for a new deferring gesture recognizer which prevents all pan and pinch gestures at (or
underneath) the web view from firing, until the web page has had a chance to prevent active
"touchmove" events. Note that we're able to forgo the need for separate "immediately resettable"
and "delayed resettable" deferring gestures here, since the only native gestures that can be
prevented by preventing "touchmove" events are pan and pinch gestures, which immediately reset in
the first place.

(-[WKContentView setUpInteraction]):
(-[WKContentView _webTouchEventsRecognized:]):
(-[WKContentView deferringGestures]):
(-[WKContentView _doneDeferringTouchMove:]):
(-[WKContentView deferringGestureRecognizer:didEndTouchesWithEvent:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::cancelCurrentInteractionInformationRequest):
(WebKit::WebPage::cancelGesturesBlockedOnSynchronousReplies): Deleted.
(WebKit::WebPage::touchEventSync): Deleted.

Delete more code (and hacks) around synchronous touch event dispatch.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/252909@main
